### ISSUE TITLE
Make gh-pages branch protected

### ISSUE THREAD

**Issue by Andarist** [redux-saga/redux-saga#1252]
I've deleted by accident today `gh-pages` branch which contains the documentation and automatically our docs site went down.

Would be desirable to have this branch protected, but this mean we can't force push to it and that's what currently [our publication script](https://github.com/redux-saga/redux-saga/blob/1877d2ea3173e61b01407edf6475f25116b386df/package.json#L42) does. It needs to be tweaked to comit the content and push it so we could make `gh-pages` protected.

**Comment by lukePeavey** [comment]
@Andarist Is this just a matter of removing the `--force` flag from the `push` command, or is it more involved than that?

**Comment by Andarist** [comment]
I imagine it requires removing the `--force` flag and changing the behaviour of the script to commit new content each time

**Comment by lukePeavey** [comment]
Thanks for the response.  I wasn't able to figure out how to implement this.

**Comment by Andarist** [comment]
I got an advice from a friend who knows git way better than me, that it would probably require:
- `GIT_WORK_TREE=docs/ git checkout gh-pages` - which should unpack `gh-pages` into `docs`
- `GIT_WORK_TREE=docs/ git commit -am 'update book'`
- and then just do a push

This recipe is probably missing something, but at least it is a suggestion what would need to be explored to implement this.

**Comment by lukePeavey** [comment]
Ok, that definitely helps! At first glance I didn't fully understand what was going on in the `docs:publish` script, but after breaking it down and doing some reading about gitbooks deployment, i think i have a better handle on it. 

Im going to some work on this and see if I can figure it out.

**Comment by lukePeavey** [comment]
So far, I haven't been able to get this working with the `GIT_WORK_TREE` approach. I did find another possible solution using the [gh-pages module](https://www.npmjs.com/package/gh-pages). I came across it in [this article](https://medium.com/@gpbl/how-to-use-gitbook-to-publish-docs-for-your-open-source-npm-packages-465dd8d5bfba).

It accomplishes the same thing as the current publication script, but doesn't use `--force` push to overwrite the `gh-pages` branch each time. 

From the documentation
> It will create a temporary clone of the current repository, create a gh-pages branch if one doesn't already exist, copy over all files from the base path, or only those that match patterns from the optional src configuration, commit all changes, and push to the origin remote.
>
>If a gh-pages branch already exists, it will be updated with all commits from the remote before adding any commits from the provided src files.

The publication script would be changed to
```sh
# Install GitBook dependencies and build the documentation site to the _book dir
npm run docs:clean && npm run docs:build
# Copy the custom domain file to the _book dir
cp CNAME _book
# Commit and push the contents of the _book dir to the gh-pages branch on the remote
gh-pages --dist=_book --repo="git@github.com:redux-saga/redux-saga.git" -m "Update book"
```
I've tested this on a [clone of the redux-saga repo](https://github.com/lukePeavey/gitbook-deploy-script-test) and it seems to be working. If you think its a possibility, i can explore it further.

**Comment by Andarist** [comment]
It certainly is a possibility, I don't mind using any package if it works :)

**Comment by lukePeavey** [comment]
I went ahead and submitted a PR, hope thats ok.

**Comment by Andarist** [comment]
Sure thing! Gonna review, hopefully soon :)

**Comment by Andarist** [comment]
closed by #1290 thanks to @lukePeavey
